src/types/bytearray.rs

```diff
@@ -71,6 +71,24 @@ impl PyByteArray {
         }
     }
 
+    /// Creates a new Python bytearray object from a raw pointer and length.
+    ///
+    /// Panics if out of memory.
+    ///
+    /// # Safety
+    ///
+    /// This function dereferences the raw pointer `ptr` as the
+    /// leading pointer of a slice of length `len`. [As with
+    /// `std::slice::from_raw_parts`, this is
+    /// unsafe](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety).
+    /// If `ptr` is null, the contents of the bytearray object are uninitialized.
+    pub unsafe fn from_ptr(py: Python<'_>, ptr: *const u8, len: usize) -> &PyByteArray {
```

Comment:
I understand why the C API needs to be unsafe, but why don't we wrap this like

```rust
pub fn from_slice(py: Python<'_>, slice: &[u8]) -> &PyByteArray { ... }
```

if the pointer is referenced that way in any case?

As it is written it seems to offer very little that the FFI entry point does not.

Reply:
(Or `slice: Option<&[u8]>` if we want to support the nullable case.)

Reply:
This would also better communicate what the function does. Does it copy the contents of the slice or does it just reference the pointer? I'm guessing the former but I'm not sure.

Reply:
https://github.com/python/cpython/blob/37bd8726b87c1c41f58cef031f5b6d5f900d265b/Objects/bytearrayobject.c#L141

It calls `memcpy` with the pointer and size.

Reply:
> As it is written it seems to offer very little that the FFI entry point does not.

Yes, it's just more convenient.

> ```rust
> pub fn from_slice(py: Python<'_>, slice: &[u8]) -> &PyByteArray { ... }
> ```

I'm fine with also adding this, the nullable case is unsafe so I think it's better for `from_slice` to just take `&[u8]` and use `from_ptr` for the nullable case.

Reply:
Is it problematic to have a way to read uninitialised memory from the nullable case? It seems a little scary to me, but I think because of the FFI boundary in practice this won't trigger UB and instead just gets treatedas arbitrary bytes slice.

What does it look like to print the uninitialised bytearray in Python? Might be a bit too easy to accidentally leak sensitive data if the buffer just gets printed.

We could have a `PyByteArray::zeroed` safe constructor, maybe?